G1ANT.Addon.Net/G1ANT.Addon.Net/Commands/MailImapCommand.cs

```diff
@@ -0,0 +1,137 @@
+
+            ClientConnection(client, credentials, uri, timeout, markAsRead == false);
+
+            if (client.IsConnected && client.IsAuthenticated)
+            {
+                var messages = ReceiveMesssages(client, arguments);
+
+                SendMessageListToScripter(client, arguments, messages);
+
+                if (markAsRead)
+                {
+                    MarkMessagesAsRead(client, messages);
+                }
+            }
+        }
+
+        private void ClientConnection(ImapClient client, NetworkCredential credentials, Uri uri, int timeout, bool readOnly)
+        {
+            client.Timeout = timeout;
+            client.ServerCertificateValidationCallback = (s, c, h, e) => true; //HACK try to Repair & Remove
+            client.Connect(uri);
+            client.Authenticate(credentials);
+            client.Inbox.Open(readOnly ? FolderAccess.ReadOnly : FolderAccess.ReadWrite);
+            client.Inbox.Subscribe();
+        }
+
+        private void SendMessageListToScripter(ImapClient client, Arguments arguments, List<IMessageSummary> messages)
+        {
+            var messageList = new ListStructure();
+
+            foreach (var message in messages)
+            {
+                var messageWithFolder = new SimplifiedMessageSummary(message as MessageSummary, client.Inbox);
+                var structure = new MailStructure(messageWithFolder, null, null);
+                messageList.AddItem(structure);
+            }
+
+            Scripter.Variables.SetVariableValue(arguments.Result.Value, messageList);
+        }
+
+
+        private List<IMessageSummary> ReceiveMesssages(ImapClient client, Arguments arguments)
+        {
+            var allMessages = client.Inbox.Fetch(0, -1, MessageSummaryItems.All | MessageSummaryItems.Body | MessageSummaryItems.BodyStructure | MessageSummaryItems.UniqueId).ToList();
+            var onlyUnread = arguments.OnlyUnreadMessages.Value;
+            var since = arguments.SinceDate.Value;
+            var to = arguments.ToDate.Value;
+            return SelectMessages(allMessages, onlyUnread, since, to);
+        }
+
+        private void MarkMessagesAsRead(ImapClient client, List<IMessageSummary> messages)
+        {
+            foreach (IMessageSummary message in messages)
+            {
+                client.Inbox.SetFlags(message.UniqueId, MessageFlags.Seen, true);
+            }
+        }
+
+        private List<IMessageSummary> SelectMessages(List<IMessageSummary> messages, bool onlyUnRead, DateTime sinceDate,
+                                                     DateTime toDate)
+        {
+            Func<IMessageSummary, bool> isUnread = m => m.Flags != null && m.Flags.Value.HasFlag(MessageFlags.Seen) == false;
+            var relevantMessages = onlyUnRead ? messages.Where(isUnread).ToList() : messages;
+
+            relevantMessages = relevantMessages.Where(m => m.Date >= sinceDate && m.Date <= toDate).ToList();
```

Comment:
`onlyUnRead == true` => you read all unread messages, then you filter and read them again by date. You possible waste resources (bandwidth and wait time).